### PR TITLE
homed: avoid double encryption

### PR DESCRIPTION
**Pull request by poettering**
This adds an automatism: if /home is already encrypted let's not use LUKS as storage, because then we'd put an encrypted container onto an encrypted backend, and we really should avoid double encryption by default. After all, if the admin decided to encrypt everything anyway, and accept that the keys for that must be shared by all local users there's no point in second guessing that and adding another layer of encryption that just makes things slower.

Also, make the the default storage an fs choice configurable in homed.conf

### CODE REVIEW

File: man/rules/meson.build [hunk 1]
```diff
@@ -718,7 +719,11 @@ manpages = [
  ['sd_machine_get_class', '3', ['sd_machine_get_ifindices'], ''],
  ['sd_notify',
   '3',
-  ['sd_notifyf', 'sd_pid_notify', 'sd_pid_notify_with_fds', 'sd_pid_notifyf', 'sd_notify_barrier'],
+  ['sd_notify_barrier',
+   'sd_notifyf',
+   'sd_pid_notify',
+   'sd_pid_notify_with_fds',
+   'sd_pid_notifyf'],
```
Comment (keszybz) [inline man/rules/meson.build]: Not your fault obviously, but doesn't belong here. Maybe split this out into a separate commit?

File: tools/make-directive-index.py [hunk 2]
```diff
@@ -62,6 +62,14 @@
                 <variablelist id='efi-variables' />
         </refsect1>
 
+        <refsect1>
+                <title>Home Area/User Account directives</title>
+
+                <para>Directives for configuring home areas and user accounts via <citerefentry><refentrytitle>systemd-homed.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>.</para>
```
Comment (keszybz) [inline tools/make-directive-index.py]: Line break?

File: man/homed.conf.xml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    <variablelist class='home-directives'>
+
+      <varlistentry>
+        <term><varname>DefaultStorage=</varname></term>
+        <listitem><para>The default storage to use for home areas. Takes one of <literal>luks</literal>,
+        <literal>fscrypt</literal>, <literal>directory</literal>, <literal>subvolume</literal>,
+        <literal>cifs</literal>. For details about these options, see
+        <citerefentry><refentrytitle>homectl</refentrytitle><manvolnum>1</manvolnum></citerefentry>. If not
+        configured or assigned the empty string the default storage is automatically determined: if not
```
Comment (keszybz) [inline man/homed.conf.xml]: Please put a comma after "the empty string".

File: man/homed.conf.xml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        running in a container enviroment and <filename>/home/</filename> is not itself encrypted defaults to
```
Comment (keszybz) [inline man/homed.conf.xml]: Here too, comma after "is not itself encrypted".

File: src/test/test-fs-util.c [hunk 5]
```diff
@@ -846,6 +846,26 @@ static void test_chmod_and_chown_unsafe(void) {
...
+        r = path_is_encrypted(p);
+        assert_se(r >= 0);
+
+        printf("%s encrypted: %s\n", p, yes_no(r));
+
+        assert_se(expect < 0 || ((r > 0) == (expect > 0)));
+}
+
+static void test_path_is_encrypted(void) {
+        test_path_is_encrypted_one("/home", -1);
```
Comment (keszybz) [inline src/test/test-fs-util.c:863]: Please add a header as in other functions here.

File: src/test/test-fs-util.c [hunk 6]
```diff
@@ -846,6 +846,26 @@ static void test_chmod_and_chown_unsafe(void) {
...
+        test_path_is_encrypted_one("/var", -1);
+        test_path_is_encrypted_one("/", -1);
+        test_path_is_encrypted_one("/proc", false);
+        test_path_is_encrypted_one("/sys", false);
+        test_path_is_encrypted_one("/dev", false);
```
Comment (keszybz) [inline src/test/test-fs-util.c:868]: Hmm, this doesn't seem to work at all:
```console
$ build/test-fs-util
...
/home encrypted: no
/var encrypted: no
/ encrypted: no
/proc encrypted: no
/sys encrypted: no
/dev encrypted: no
```
and the same with `sudo`. I have full-disk encryption with LVM. I guess this is more complicated to discover, but it is a very popular setup and inability to detect this limits the usefulness of this check significantly.

File: src/home/homework.c [hunk 7]
```diff
@@ -836,9 +837,68 @@ static int user_record_compile_effective_passwords(
...
+        int r;
+
+        assert(ret);
+
+        /* homed tells us via an environment variable which default storage to use */
+        e = getenv("SYSTEMD_HOME_DEFAULT_STORAGE");
+        if (e) {
+                storage = user_storage_from_string(e);
+                if (storage < 0)
+                        log_warning("$SYSTEMD_HOME_DEFAULT_STORAGE set to invalid storage type, ignoring: %s", e);
```
Comment (keszybz) [inline src/home/homework.c:878]: Dunno, this would be an internal programming error. Shouldn't this result in hard failure?

File: src/home/homework.c [hunk 8]
```diff
@@ -836,9 +837,68 @@ static int user_record_compile_effective_passwords(
...
+                else {
+                        log_info("Using configured default storage '%s'.", user_storage_to_string(storage));
+                        *ret = storage;
+                        return 0;
+                }
+        }
+
+        /* When neither user nor admin specified the storage type to use, fix it to be LUKS -- unless we run
```
Comment (keszybz) [inline src/home/homework.c]: —

File: src/home/homed-conf.c [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+                *s = mfree(s);
+        else {
+                if (!supported_fstype(rvalue)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0, "Unsupported file system, ignoring: %s", rvalue);
+                        return 0;
+                }
+
+                if (free_and_strdup(s, rvalue) < 0)
+                        return log_oom();
+        }
```
Comment (keszybz) [inline src/home/homed-conf.c:47]:
```C
if (!isempty(rvalue) && !supported_fstype(rvalue)) {
                        log_syntax(unit, LOG_ERR, filename, line, 0, "Unsupported file system, ignoring: %s", rvalue);
                        return 0;
}

return free_and_strdup_warn(s, empty_to_null(rvalue));
```
?

File: man/homectl.xml [hunk 10]
```diff
@@ -504,9 +504,9 @@
         <listitem><para>Selects the storage mechanism to use for this home directory. Takes one of
         <literal>luks</literal>, <literal>fscrypt</literal>, <literal>directory</literal>,
         <literal>subvolume</literal>, <literal>cifs</literal>. For details about these mechanisms, see
-        above. If a new home directory is created and the storage type is not specifically specified defaults
-        to <literal>luks</literal> if supported, <literal>subvolume</literal> as first fallback if supported,
-        and <literal>directory</literal> if not.</para></listitem>
+        above. If a new home directory is created and the storage type is not specifically specified
```
Comment (keszybz) [inline man/homectl.xml]: Comma after "specified".

File: src/test/test-fs-util.c (hunk 6 above)
Comment (poettering) [reply src/test/test-fs-util.c:868]: so this would be relatively easy to get working. I have no LVM and didn't want to set it up, that's all. I couldn't test it without...

When determining a block device is not encrypted, path_is_encrypted() should iterate through /sys/dev/block/%u:%u/slaves/, which lists all backing devices of the block device if it exists. It should then just run the original test again on all listed block devices, and return true if all of them are encrypted. As extra safety precaution it might make sense to not descend endlessly though, but put a limit on it.

But maybe let's do that in a separate PR.

Comment (keszybz) [reply src/test/test-fs-util.c:868]: OK.

File: src/home/homework.c (hunk 7 above)
Comment (poettering) [reply src/home/homework.c:878]: well, users might set the env var in the service file. the config file will override it then if it is set. but if it isn't set it will just be propagated down the process tree and homework will honour it. While I wouldn't advertise that this works I'd still cater enough for it to work sensibly...

File: man/rules/meson.build (hunk 1 above)
Comment (poettering) [reply man/rules/meson.build]: I split this out into #15743, because it has nothing to do with the rest of this PR.

**Comment by poettering** [comment]
force pushed a new version. No changes but the suggested ones, but see comments above.

File: src/test/test-fs-util.c (hunk 6 above)
Comment (poettering) [reply src/test/test-fs-util.c:868]: I added (lightly tested) support for such complex storage in #15744, ptal!
